Fix App Bundle/Release build missing index.android.bundle with  gradle plugin 4.1.0+/gradle 6.5

tomoima525 opened the pull request:
## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

- This fix resolves https://github.com/facebook/react-native/issues/29398
- After updating gradle to 6.5+ and android gradle plugin to 4.1.0+(which is recommended in the latest Android Studio 4.1), Running `:app:assembleRelease` or `:app:bundleRelease` will not contain `index.android.bundle` in Apk/AAB. It will be included when the command executed twice.
<img width="949" alt="Screen Shot 2020-10-17 at 11 32 43 PM" src="https://user-images.githubusercontent.com/6277118/96360808-38165c00-10d5-11eb-8b6e-f098517a24c7.png">

- This is caused by the task ordering update introduced in gradle plugin 4.1.0+/gradle 6.5. `mergeResources` task runs before `currentAssetsCopyTask`(copying the bundle asset file to intermediate output directory) which causes generated Apk/AAB not including the bundle file.
- The fix ensures mergeResources task runs after currentAssetsCopyTask


## Changelog

[Android] [Fixed] - Fix App Bundle/Release build missing index.android.bundle with gradle plugin 4.1.0/gradle 6.5

## Test Plan
- Reproducible repository https://github.com/tomoima525/android_build_test_rn 
  -  This project is generated with `create-react-native-app` and updated Gradle version to 6.5 and com.android.tools.build:gradle plugin to 4.1
- Run `./gradlew clean :app:assembleRelease` and `./gradlew clean :app:bundleRelease` => reproduces the issue
- After adding the fix above and run `./gradlew clean :app:assembleRelease` and `./gradlew clean :app:bundleRelease` => The issue is resolved

- Also confirmed the build works properly with android gradle plugin `3.5.3` and `gradle 6.2`(the default value of `create-react-native-app`)

emeraldsanto commented:
I've also been experiencing this since updating this morning, reverting to 4.0.1 and 6.3 works until this PR is merged :)

tomoima525 commented:
Just rebased so that the commit message is more descriptive

tomoima525 commented:
I have the reproducible case and fixed case with this patch

https://github.com/tomoima525/android_build_test_rn

## Reproducible case

To reproduce the issue, run the command below

```
$ cd android
$ ./gradlew clean :app:assembleRelease // or ./gradlew clean :app:releaseBundle
```

app-release.apk will be generated but it does not have index.android.bundle build

## Fixed case

I added the fix in https://github.com/facebook/react-native/pull/30177 . Running the command below will apply the patch to `react.gradle` and you can see the issue is resolved

```
$ yarn add_patch
$ cd android
$ ./gradlew clean :app:assembleRelease // or ./gradlew clean :app:releaseBundle
```

(The script is tested on Mac. If the copying does not work on Linux, try directly copying the `react-patch.gradle` to `node_modules/react-native/react.gradle`)

mikehardy commented:
Confirmed for me this fixes the error it says it fixes for RN0.63.3 and android gradle plugin 4.1.0

cat `patches/react-native+0.63.3.patch`

```
diff --git a/node_modules/react-native/react.gradle b/node_modules/react-native/react.gradle
index 6441d93..c28cbd4 100644
--- a/node_modules/react-native/react.gradle
+++ b/node_modules/react-native/react.gradle
@@ -302,6 +302,11 @@ afterEvaluate {
             enabled(currentBundleTask.enabled)
         }
 
+        // mergeResources task runs before the bundle file is copied to the intermediate asset directory from Android plugin 4.1+.
+        // This ensures to copy the bundle file before mergeResources task starts
+        def mergeResourcesTask = tasks.findByName("merge${targetName}Resources")
+        mergeResourcesTask.dependsOn(currentAssetsCopyTask)
+
         packageTask.dependsOn(currentAssetsCopyTask)
         if (buildPreBundleTask != null) {
             buildPreBundleTask.dependsOn(currentAssetsCopyTask)
```

cpojer reviewed (approved):
Thanks for your contribution! 👍

John-Konordo commented:
Development machine OS: MacOS X
IDE: Ms VSCODE
"react": "^17.0.1",
"react-native": "^0.63.4",
gradle plugin 4.1.1+ / gradle 6.5-all (distributionUrl=https\://services.gradle.org/distributions/gradle-6.5-all.zip)

I found a workarround to publish my apk to play console and don' t get the error index.android.bundle missing

In root directory of your project run the following commands
cd ./android/
./gradlew clean
rm -rf ~/.gradle/caches/*
cd ..
npx react-native bundle --platform android --dev false --entry-file index.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res
rm -rf ./android/app/src/main/res/drawable-*
rm -rf ./android/app/src/main/res/raw
cd ./android/
./gradlew assembleRelease

Install my apk to any device and works well and upload the apk to play console and the error dissapear too!

pixelknitter commented:
> Confirmed for me this fixes the error it says it fixes for RN0.63.3 and android gradle plugin 4.1.0
> 
> Here's a patch-package compatible thing you can cut-and-paste
> 
> cat `patches/react-native+0.63.3.patch`
> 
> ```
> diff --git a/node_modules/react-native/react.gradle b/node_modules/react-native/react.gradle
> index 6441d93..c28cbd4 100644
> --- a/node_modules/react-native/react.gradle
> +++ b/node_modules/react-native/react.gradle
> @@ -302,6 +302,11 @@ afterEvaluate {
>              enabled(currentBundleTask.enabled)
>          }
>  
> +        // mergeResources task runs before the bundle file is copied to the intermediate asset directory from Android plugin 4.1+.
> +        // This ensures to copy the bundle file before mergeResources task starts
> +        def mergeResourcesTask = tasks.findByName("merge${targetName}Resources")
> +        mergeResourcesTask.dependsOn(currentAssetsCopyTask)
> +
>          packageTask.dependsOn(currentAssetsCopyTask)
>          if (buildPreBundleTask != null) {
>              buildPreBundleTask.dependsOn(currentAssetsCopyTask)
> ```

@mikehardy For those discovering this solution, I thought I'd figure out exact instructions to patch from 0.63.3 as my project is not ready to migrate to 0.64; this seems like a good option. I've not used `patch-package` before, running `patch-package patches/react-native+0.63.3.patch` would require me to make the changes to the `node_modules` myself first, correct?

mikehardy commented:
@pixelknitter - nope, `yarn add patch-package && mkdir patches` drop the patch file into there (named correctly! that's how patch-package figures out what package etc) then `yarn patch-package` should do it. Or all the equivalent-different-but-same npm commands.

If you are making a *new* patch (like how I generated this file) to fix some bug where you are the first person to find+fix, then you hack around in node_modules by hand yes, and run `yarn patch-package <package name>` and it will create a new patch for you with your new change in it for distribution like this patch here

pixelknitter commented:
> @pixelknitter - nope, `yarn add patch-package && mkdir patches` drop the patch file into there (named correctly! that's how patch-package figures out what package etc) then `yarn patch-package` should do it. Or all the equivalent-different-but-same npm commands.
> 
> If you are making a _new_ patch (like how I generated this file) to fix some bug where you are the first person to find+fix, then you hack around in node_modules by hand yes, and run `yarn patch-package <package name>` and it will create a new patch for you with your new change in it for distribution like this patch here

Thanks for the detailed response!

mouhsnimohamed commented:
if you'r using proguard add an exception to android/app/proguard-rules.pro:

 -keep class com.mypackage.BuildConfig { *; }

com.mypackage should match the package value in your app/src/main/AndroidManifest.xml file.

xczaixian commented:
> Confirmed for me this fixes the error it says it fixes for RN0.63.3 and android gradle plugin 4.1.0
> 
> Here's a patch-package compatible thing you can cut-and-paste
> 
> cat `patches/react-native+0.63.3.patch`
> 
> ```
> diff --git a/node_modules/react-native/react.gradle b/node_modules/react-native/react.gradle
> index 6441d93..c28cbd4 100644
> --- a/node_modules/react-native/react.gradle
> +++ b/node_modules/react-native/react.gradle
> @@ -302,6 +302,11 @@ afterEvaluate {
>              enabled(currentBundleTask.enabled)
>          }
>  
> +        // mergeResources task runs before the bundle file is copied to the intermediate asset directory from Android plugin 4.1+.
> +        // This ensures to copy the bundle file before mergeResources task starts
> +        def mergeResourcesTask = tasks.findByName("merge${targetName}Resources")
> +        mergeResourcesTask.dependsOn(currentAssetsCopyTask)
> +
>          packageTask.dependsOn(currentAssetsCopyTask)
>          if (buildPreBundleTask != null) {
>              buildPreBundleTask.dependsOn(currentAssetsCopyTask)
> ```

thank you, but it's not work for me.
i supplement code below to make it work.
`
def compressAssetsTask = tasks.findByName("compress${targetName}Assets")
compressAssetsTask.dependsOn(currentAssetsCopyTask)
`

mikehardy commented:
@xczaixian this is no longer a problem with current versions (0.64.1) - I'm not carrying this patch any longer as it has been integrated and released. Better to update and get the real fix

9christian9 commented:
Ciao,
I'm writing for all the people who, like me, have experienced this problem.

**Affected Versions**
- React Native: 0.67.x
- Gradle 7.2.x - 7.3.x


**The common error**
> Unable to load script. Make sure you're either running a Metro server (run 'react-native start') or that your bundle 'index.android.bundle' is packaged correctly for release.

In short, index.android.bundle is missing


**The problem**
I found that the problem is in a native react gradle task, to be precise in the copy task called `currentAssetsCopyTask` located in `react.gradle`file.
As I understand, this "task" is responsible for copying the `assets` and in particular `index.android.bundle` and pasting them into the temporary directories used to generate the APK.


**The solution**
I created a repo that with a js script solves the problem before building the android app
[https://github.com/9christian9/FIX-Android-productFlavors-RN0.67-gradle7.2-7.3](https://github.com/9christian9/FIX-Android-productFlavors-RN0.67-gradle7.2-7.3)


I hope this will help.
With ♥ by Christian
